docker/docker_image_dest.go

```diff
@@ -131,15 +132,49 @@ func (d *dockerImageDestination) HasThreadSafePutBlob() bool {
 // to any other readers for download using the supplied digest.
 // If stream.Read() at any time, ESPECIALLY at end of input, returns an error, PutBlob MUST 1) fail, and 2) delete any data stored so far.
 func (d *dockerImageDestination) PutBlob(ctx context.Context, stream io.Reader, inputInfo types.BlobInfo, cache types.BlobInfoCache, isConfig bool) (types.BlobInfo, error) {
+	digester, stream := putblobdigest.DigestIfCanonicalUnknown(stream, inputInfo)
+	sizeCounter := &sizeCounter{}
+	stream = io.TeeReader(stream, sizeCounter)
+
+	// If requested, precompute the blob digest to prevent uploading layers that already exist on the registry.
+	// This functionality is particularly useful when BlobInfoCache has not been populated with compressed digests,
+	// the source blob is uncompressed, and the destination blob is being compressed "on the fly".
+	if inputInfo.Digest == "" && d.c.sys.DockerRegistryPushPrecomputeDigests {
+		logrus.Debugf("Precomputing digest layer for %s", reference.Path(d.ref.ref))
+		diskBlob, err := ioutil.TempFile(tmpdir.TemporaryDirectoryForBigFiles(d.c.sys), "docker-layer-blob")
+		if err != nil {
+			return types.BlobInfo{}, errors.Wrap(err, "creating temporary on-disk layer")
```

Comment:
Can you turn `errors.Wrap(err, "")` into `fmt.Errorf("...: %w", err)`?  Old code is still using the errors package but we want new code to use the native error handling in Go.